"use strict";(self.webpackChunkcypress_docusaurus_zh=self.webpackChunkcypress_docusaurus_zh||[]).push([[6141],{5055:(e,n,t)=>{t.r(n),t.d(n,{assets:()=>c,contentTitle:()=>a,default:()=>h,frontMatter:()=>o,metadata:()=>r,toc:()=>l});var s=t(5893),i=t(1151);const o={title:"Retry-ability",sidebar_position:50},a=void 0,r={id:"app/core-concepts/retry-ability",title:"Retry-ability",description:"What you'll learn",source:"@site/docs/app/core-concepts/retry-ability.mdx",sourceDirName:"app/core-concepts",slug:"/app/core-concepts/retry-ability",permalink:"/app/core-concepts/retry-ability",draft:!1,unlisted:!1,editUrl:"https://github.com/cypress-io/cypress-documentation/tree/main/docs/app/core-concepts/retry-ability.mdx",tags:[],version:"current",lastUpdatedAt:1732009455,formattedLastUpdatedAt:"Nov 19, 2024",sidebarPosition:50,frontMatter:{title:"Retry-ability",sidebar_position:50},sidebar:"app",previous:{title:"Test Isolation",permalink:"/app/core-concepts/test-isolation"},next:{title:"Open Mode",permalink:"/app/core-concepts/open-mode"}},c={},l=[{value:"<Icon></Icon> What you&#39;ll learn",id:"What-youll-learn",level:5},{value:"Commands, Queries and Assertions",id:"Commands-Queries-and-Assertions",level:2},{value:"Multiple assertions",id:"Multiple-assertions",level:2},{value:"Implicit Assertions",id:"Implicit-Assertions",level:2},{value:"Timeouts",id:"Timeouts",level:2},{value:"Increase time to retry",id:"Increase-time-to-retry",level:3},{value:"Disable retry",id:"Disable-retry",level:3},{value:"Only queries are retried",id:"Only-queries-are-retried",level:2},{value:"Actions should be at the end of chains, not the middle",id:"Actions-should-be-at-the-end-of-chains-not-the-middle",level:3},{value:"<Icon></Icon> Incorrectly chaining commands",id:"Incorrectly-chaining-commands",level:4},{value:"<Icon></Icon> Correctly ending chains after an action",id:"Correctly-ending-chains-after-an-action",level:4},{value:"<Icon></Icon> Incorrectly checking if the stub was called",id:"Incorrectly-checking-if-the-stub-was-called",level:4},{value:"<Icon></Icon> Correctly waiting for the stub to be called",id:"Correctly-waiting-for-the-stub-to-be-called",level:4},{value:"Use <code>.should()</code> with a callback",id:"Use-should-with-a-callback",level:3},{value:"<Icon></Icon> Incorrectly waiting for values",id:"Incorrectly-waiting-for-values",level:3},{value:"<Icon></Icon> Correctly waiting for values",id:"Correctly-waiting-for-values",level:4},{value:"See also",id:"See-also",level:2}];function d(e){const n={a:"a",admonition:"admonition",code:"code",em:"em",h2:"h2",h3:"h3",h4:"h4",h5:"h5",li:"li",p:"p",pre:"pre",strong:"strong",ul:"ul",...(0,i.a)(),...e.components},{DocsImage:t,Icon:o}=n;return t||u("DocsImage",!0),o||u("Icon",!0),(0,s.jsxs)(s.Fragment,{children:[(0,s.jsxs)(n.admonition,{type:"info",children:[(0,s.jsxs)(n.h5,{id:"What-youll-learn",children:[(0,s.jsx)(o,{name:"question-circle",color:"#4BBFD2"})," What you'll learn"]}),(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsx)(n.li,{children:"The difference between commands, queries, and assertions in Cypress"}),"\n",(0,s.jsx)(n.li,{children:"How Cypress retries multiple assertions"}),"\n",(0,s.jsx)(n.li,{children:"How to increase the time Cypress retries commands"}),"\n",(0,s.jsx)(n.li,{children:"How Cypress retries queries with examples"}),"\n"]})]}),"\n",(0,s.jsx)(n.p,{children:"A core feature of Cypress that assists with testing dynamic web applications is\nretry-ability. Like a good transmission in a car, it usually works without you\nnoticing it. But understanding how it works will help you write faster tests\nwith fewer run-time surprises."}),"\n",(0,s.jsx)(n.admonition,{type:"tip",children:(0,s.jsxs)(n.p,{children:["If you are looking to retry tests a configured number of times when the test\nfails, check out our guide on ",(0,s.jsx)(n.a,{href:"/app/guides/test-retries",children:"Test Retries"}),"."]})}),"\n",(0,s.jsx)(n.h2,{id:"Commands-Queries-and-Assertions",children:"Commands, Queries and Assertions"}),"\n",(0,s.jsxs)(n.p,{children:["While all methods you chain off of ",(0,s.jsx)(n.code,{children:"cy"})," in your Cypress tests are commands, it's\nimportant to understand the different rules by which they operate."]}),"\n",(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsxs)(n.li,{children:[(0,s.jsx)(n.strong,{children:"Queries"})," link up, retrying the entire chain together."]}),"\n",(0,s.jsxs)(n.li,{children:[(0,s.jsx)(n.strong,{children:"Assertions"})," are a type of query that's specially displayed in the command\nlog."]}),"\n",(0,s.jsxs)(n.li,{children:[(0,s.jsx)(n.strong,{children:"Non-queries"})," only execute once."]}),"\n"]}),"\n",(0,s.jsx)(n.p,{children:"For example, there are 5 queries, an action, and 2 assertions in\nthe test below."}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-javascript",children:"it('creates an item', () => {\n  // Non-query commands only execute once.\n  cy.visit('/')\n\n  // The .focused() query and .should() assertion link together,\n  // rerunning until the currently focused element has\n  // the 'new-todo' class\n  cy.focused().should('have.class', 'new-todo')\n\n  // The queries .get() and .find()\n  // link together, forming the subject for\n  // the non-query `.type()`.\n  cy.get('.header').find('.new-todo').type('todo A{enter}')\n\n  // Two queries and an assertion chained together\n  cy.get('.todoapp').find('.todo-list li').should('have.length', 1)\n})\n"})}),"\n",(0,s.jsxs)(n.p,{children:["The ",(0,s.jsx)(n.a,{href:"/app/core-concepts/open-mode#Command-Log",children:"Command Log"})," shows all\ncommands regardless of types, with passing assertions showing in green."]}),"\n",(0,s.jsx)(t,{src:"/img/app/retry-ability/command-assertions.png",alt:"Cypress tests showing commands and assertions"}),"\n",(0,s.jsx)(n.p,{children:"Let's look at the last chain of commands:"}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-javascript",children:"cy.get('.todoapp') // query\n  .find('.todo-list li') // query\n  .should('have.length', 1) // assertion\n"})}),"\n",(0,s.jsxs)(n.p,{children:["Because nothing is synchronous in modern web applications, Cypress can't query\nall the DOM elements matching ",(0,s.jsx)(n.code,{children:".todo-list li"})," and check if there is exactly one\nof them. There are many examples of why this would not work well."]}),"\n",(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsx)(n.li,{children:"What if the application has not updated the DOM by the time these commands\nrun?"}),"\n",(0,s.jsx)(n.li,{children:"What if the application is waiting for its back end to respond before\npopulating the DOM element?"}),"\n",(0,s.jsx)(n.li,{children:"What if the application does some intensive computation before showing the\nresults in the DOM?"}),"\n"]}),"\n",(0,s.jsxs)(n.p,{children:["Thus ",(0,s.jsx)(n.a,{href:"/api/commands/get",children:(0,s.jsx)(n.code,{children:"cy.get"})})," and ",(0,s.jsx)(n.a,{href:"/api/commands/find",children:(0,s.jsx)(n.code,{children:"cy.find()"})})," have to\nbe smarter and expect the application to potentially update. ",(0,s.jsx)(n.code,{children:"cy.get()"})," queries\nthe application's DOM, finds the elements that match the selector, and then\npasses them to ",(0,s.jsx)(n.code,{children:".find('.todo-list li')"}),". ",(0,s.jsx)(n.code,{children:".find()"})," locates a new set of\nelements, and passes them to ",(0,s.jsx)(n.code,{children:".should()"}),". ",(0,s.jsx)(n.code,{children:".should()"})," then asserts on the list\nof found elements (in our case, that it has a length of 1)."]}),"\n",(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsxs)(n.li,{children:["\u2705 If the assertion passes, then ",(0,s.jsx)(n.code,{children:".should()"})," finishes successfully."]}),"\n",(0,s.jsxs)(n.li,{children:["\ud83d\udea8 If the assertion fails, then Cypress will requery the application's DOM\nagain - starting from the top of the chain of linked queries. It will look for\nelements that match ",(0,s.jsx)(n.code,{children:".get().find()"})," again, and re-run the assertion. If the\nassertion still fails, Cypress continues retrying until the timeout is\nreached."]}),"\n"]}),"\n",(0,s.jsx)(n.p,{children:"Retry-ability allows the test to complete each command as soon as the assertion\npasses, without hard-coding waits. If your application takes a few milliseconds\nor even seconds to render each DOM element - no big deal, the test does not have\nto change at all. For example, let's introduce an artificial delay of 3 seconds\nwhen refreshing the application's UI below in an example TodoMVC model code:"}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-javascript",children:"app.TodoModel.prototype.addTodo = function (title) {\n  this.todos = this.todos.concat({\n    id: Utils.uuid(),\n    title: title,\n    completed: false,\n  })\n\n  // let's trigger the UI to render after 3 seconds\n  setTimeout(() => {\n    this.inform()\n  }, 3000)\n}\n"})}),"\n",(0,s.jsxs)(n.p,{children:["The test still passes! ",(0,s.jsx)(n.code,{children:"cy.get('.todo-list')"})," passes immediately - the\n",(0,s.jsx)(n.code,{children:"todo-list"})," exists - but ",(0,s.jsx)(n.code,{children:".should('have.length', 1)"})," shows the spinning\nindicator, meaning Cypress is still waiting for the assertion to pass."]}),"\n",(0,s.jsx)(t,{src:"/img/app/retry-ability/retry-assertion.gif",alt:"Retrying assertion"}),"\n",(0,s.jsxs)(n.p,{children:["Within a few milliseconds after the DOM updates, the linked queries\n",(0,s.jsx)(n.code,{children:".get().find()"})," locate an element and ",(0,s.jsx)(n.code,{children:".should('have.length', 1)"})," passes."]}),"\n",(0,s.jsx)(n.h2,{id:"Multiple-assertions",children:"Multiple assertions"}),"\n",(0,s.jsx)(n.p,{children:"Queries and assertions are always executed in order, and always retry 'from the\ntop'. If you have multiple assertions, Cypress will retry until each passes\nbefore moving on to the next one."}),"\n",(0,s.jsxs)(n.p,{children:["For example, the following test has ",(0,s.jsx)(n.a,{href:"/api/commands/should",children:(0,s.jsx)(n.code,{children:".should()"})})," and\n",(0,s.jsx)(n.a,{href:"/api/commands/and",children:(0,s.jsx)(n.code,{children:".and()"})})," assertions. ",(0,s.jsx)(n.code,{children:".and()"})," is an alias of the\n",(0,s.jsx)(n.code,{children:".should()"})," command, so the second assertion is really a custom callback\nassertion in the form of the ",(0,s.jsx)(n.a,{href:"/api/commands/should#Function",children:(0,s.jsx)(n.code,{children:".should(cb)"})}),"\nfunction with 2 ",(0,s.jsx)(n.a,{href:"/app/references/assertions#Chai",children:(0,s.jsx)(n.code,{children:"expect"})}),"\nassertions inside of it."]}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-javascript",children:"it('creates two items', () => {\n  cy.visit('/')\n\n  cy.get('.new-todo').type('todo A{enter}')\n  cy.get('.new-todo').type('todo B{enter}')\n\n  cy.get('.todo-list li') // query\n    .should('have.length', 2) // assertion\n    .and(($li) => {\n      // 2 mocha assertions inside of the .and() assertion\n      expect($li.get(0).textContent, 'first item').to.equal('todo a')\n      expect($li.get(1).textContent, 'second item').to.equal('todo B')\n    })\n})\n"})}),"\n",(0,s.jsxs)(n.p,{children:["Because the first expect statement\n(",(0,s.jsx)(n.code,{children:"expect($li.get(0).textContent, 'first item').to.equal('todo a')"}),") fails, the\nsecond statement is never reached. The ",(0,s.jsx)(n.code,{children:".and()"})," command fails after timing out,\nand the Command Log correctly shows that the first encountered assertion\n",(0,s.jsx)(n.code,{children:"should('have.length', 2)"}),' passed, but\nthe "first item" assertion failed.']}),"\n",(0,s.jsx)(t,{src:"/img/app/retry-ability/second-assertion-fails.gif",alt:"Retrying multiple assertions"}),"\n",(0,s.jsx)(n.h2,{id:"Implicit-Assertions",children:"Implicit Assertions"}),"\n",(0,s.jsxs)(n.p,{children:["Often a Cypress command has built-in assertions that will cause the command to\nbe retried until they pass. For example, the ",(0,s.jsx)(n.a,{href:"/api/commands/eq",children:(0,s.jsx)(n.code,{children:".eq()"})})," query\nwill be retried even without any attached assertions until it finds an element\nwith the given index."]}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-javascript",children:"cy.get('.todo-list li') // query\n  .should('have.length', 2) // assertion\n  .eq(3) // query\n"})}),"\n",(0,s.jsx)(t,{src:"/img/app/retry-ability/eq.gif",alt:"Retrying built-in assertion"}),"\n",(0,s.jsxs)(n.p,{children:["Only queries can be retried, but most other commands still have built-in\n",(0,s.jsx)(n.em,{children:"waiting"}),' and assertions. For example, as described in the "Assertions" section\nof ',(0,s.jsx)(n.a,{href:"/api/commands/click",children:".click()"}),", the ",(0,s.jsx)(n.code,{children:"click()"})," action command waits to click\nuntil the element becomes\n",(0,s.jsx)(n.a,{href:"/app/core-concepts/interacting-with-elements#Actionability",children:"actionable"}),",\nincluding re-running the query chain leading up to it in case the page updates\nwhile we're waiting."]}),"\n",(0,s.jsx)(n.p,{children:"Cypress tries to act like a human user would using the browser."}),"\n",(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsx)(n.li,{children:"Can a user click on the element?"}),"\n",(0,s.jsx)(n.li,{children:"Is the element invisible?"}),"\n",(0,s.jsx)(n.li,{children:"Is the element behind another element?"}),"\n",(0,s.jsxs)(n.li,{children:["Does the element have the ",(0,s.jsx)(n.code,{children:"disabled"})," attribute?"]}),"\n"]}),"\n",(0,s.jsxs)(n.p,{children:[(0,s.jsx)(n.a,{href:"/app/core-concepts/interacting-with-elements",children:"Action commands"})," - such as\n",(0,s.jsx)(n.code,{children:".click()"})," - automatically wait until multiple built-in assertions like these\npass, and then it will attempt the action once."]}),"\n",(0,s.jsx)(n.h2,{id:"Timeouts",children:"Timeouts"}),"\n",(0,s.jsxs)(n.p,{children:["By default each command that retries does so for up to 4 seconds - the\n",(0,s.jsx)(n.a,{href:"/app/references/configuration#Timeouts",children:(0,s.jsx)(n.code,{children:"defaultCommandTimeout"})})," setting."]}),"\n",(0,s.jsx)(n.h3,{id:"Increase-time-to-retry",children:"Increase time to retry"}),"\n",(0,s.jsxs)(n.p,{children:["You can change the default timeout for ",(0,s.jsx)(n.em,{children:"all commands"}),". See\n",(0,s.jsx)(n.a,{href:"/app/references/configuration#Overriding-Options",children:"Configuration: Overriding Options"}),"\nfor examples of overriding this option."]}),"\n",(0,s.jsx)(n.p,{children:"For example, to set the default command timeout to 10 seconds via the command\nline:"}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-shell",children:"cypress run --config defaultCommandTimeout=10000\n"})}),"\n",(0,s.jsxs)(n.p,{children:["We do not recommend changing the command timeout globally. Instead, pass the\nindividual command's ",(0,s.jsx)(n.code,{children:"{ timeout: ms }"})," option to retry for a different period of\ntime. For example:"]}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-javascript",children:"// we've modified the timeout which affects default + added assertions\ncy.get('[data-testid=\"mobile-nav\"]', { timeout: 10000 })\n  .should('be.visible')\n  .and('contain', 'Home')\n"})}),"\n",(0,s.jsxs)(n.p,{children:["Cypress will retry for up to 10 seconds to find a visible element with\n",(0,s.jsx)(n.code,{children:"data-testid"})," attribute ",(0,s.jsx)(n.code,{children:"mobile-nav"}),' with text containing "Home". For more\nexamples, read the\n',(0,s.jsx)(n.a,{href:"/app/core-concepts/introduction-to-cypress#Timeouts",children:"Timeouts"}),' section in\nthe "Introduction to Cypress" guide.']}),"\n",(0,s.jsx)(n.h3,{id:"Disable-retry",children:"Disable retry"}),"\n",(0,s.jsxs)(n.p,{children:["Overriding the timeout to ",(0,s.jsx)(n.code,{children:"0"})," will essentially disable retrying the query, since\nit will spend 0 milliseconds retrying."]}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-javascript",children:"// check synchronously that the element does not exist (no retry)\n// for example just after a server-side render\ncy.get('[data-testid=\"ssr-error\"]', { timeout: 0 }).should('not.exist')\n"})}),"\n",(0,s.jsx)(n.h2,{id:"Only-queries-are-retried",children:"Only queries are retried"}),"\n",(0,s.jsxs)(n.p,{children:["Any command that isn't a query, such as ",(0,s.jsx)(n.code,{children:".click()"}),', will not "chain together" to\nform the subject for later commands the way queries do. Cypress will retry any\nqueries ',(0,s.jsx)(n.em,{children:"leading up to"})," a command, and retry any assertions ",(0,s.jsx)(n.em,{children:"after"})," a command,\nbut commands themselves only execute once. After they've executed, nothing\nleading up to them will retry."]}),"\n",(0,s.jsxs)(n.p,{children:["Most commands are not retried because they could potentially change the state of\nthe application under test. For example, Cypress will not retry the\n",(0,s.jsx)(n.a,{href:"/api/commands/click",children:".click()"})," action command, because it could change\nsomething in the application. After the click occurs, Cypress will also not\nre-run any queries before ",(0,s.jsx)(n.code,{children:".click()"}),"."]}),"\n",(0,s.jsx)(n.h3,{id:"Actions-should-be-at-the-end-of-chains-not-the-middle",children:"Actions should be at the end of chains, not the middle"}),"\n",(0,s.jsx)(n.p,{children:"The following test might have problems if:"}),"\n",(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsx)(n.li,{children:"Your JS framework re-rendered asynchronously"}),"\n",(0,s.jsx)(n.li,{children:"Your app code reacted to an event firing and removed the element"}),"\n"]}),"\n",(0,s.jsxs)(n.h4,{id:"Incorrectly-chaining-commands",children:[(0,s.jsx)(o,{name:"exclamation-triangle",color:"red"})," Incorrectly chaining commands"]}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-javascript",children:"cy.get('.new-todo')\n  .type('todo A{enter}') // action\n  .type('todo B{enter}') // action after another action - bad\n  .should('have.class', 'active') // assertion after an action - bad\n"})}),"\n",(0,s.jsxs)(n.h4,{id:"Correctly-ending-chains-after-an-action",children:[(0,s.jsx)(o,{name:"check-circle",color:"green"})," Correctly ending chains after an action"]}),"\n",(0,s.jsx)(n.p,{children:"To avoid these issues entirely, it is better to split up the above chain of\ncommands."}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-javascript",children:"cy.get('.new-todo').type('todo A{enter}')\ncy.get('.new-todo').type('todo B{enter}')\ncy.get('.new-todo').should('have.class', 'active')\n"})}),"\n",(0,s.jsxs)(n.p,{children:["Writing your tests in this way will help you avoid issues where the page\nrerenders in the middle of your test and Cypress loses track of which elements\nit's supposed to be operating or asserting on. Aliases -\n",(0,s.jsx)(n.a,{href:"/api/commands/as",children:(0,s.jsx)(n.code,{children:"cy.as()"})})," - can help make this pattern less intrusive."]}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-javascript",children:"cy.get('.new-todo').as('new')\n\ncy.get('@new').type('todo A{enter}')\ncy.get('@new').type('todo B{enter}')\ncy.get('@new').should('have.class', 'active')\n"})}),"\n",(0,s.jsxs)(n.admonition,{type:"caution",children:[(0,s.jsxs)(n.p,{children:["Very rarely you may want to retry a command like ",(0,s.jsx)(n.code,{children:".click()"}),". We describe one\ncase like that where the event listeners are attached to a modal popup only\nafter a delay, thus causing default events fired during ",(0,s.jsx)(n.code,{children:".click()"}),' to not\nregister. In this special case you may want to "keep clicking" until the event\nregisters, and the dialog disappears. Read about it in the\n',(0,s.jsx)(n.a,{href:"https://www.cypress.io/blog/2019/01/22/when-can-the-test-click/",children:"When Can the Test Click?"}),"\nblog post."]}),(0,s.jsx)(n.p,{children:"Because of the implicit assertions built into every command, and action commands\nin particular, you should rarely need this pattern."})]}),"\n",(0,s.jsxs)(n.p,{children:["As another example, when confirming that the button component invokes the\n",(0,s.jsx)(n.code,{children:"click"})," prop testing with the\n",(0,s.jsx)(n.a,{href:"https://github.com/cypress-io/cypress/tree/develop/npm/react",children:"cypress/react"}),"\nmounting library, the following test might or might not work:"]}),"\n",(0,s.jsxs)(n.h4,{id:"Incorrectly-checking-if-the-stub-was-called",children:[(0,s.jsx)(o,{name:"exclamation-triangle",color:"red"})," Incorrectly checking if the stub was called"]}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-js",children:"const Clicker = ({ click }) => (\n  <div>\n    <button onClick={click}>Click me</button>\n  </div>\n)\n\nit('calls the click prop twice', () => {\n  const onClick = cy.stub()\n  cy.mount(<Clicker click={onClick} />)\n  cy.get('button')\n    .click()\n    .click()\n    .then(() => {\n      // works in this case, but not recommended\n      // because .click() and .then() do not retry\n      expect(onClick).to.be.calledTwice\n    })\n})\n"})}),"\n",(0,s.jsxs)(n.p,{children:["The above example will fail if the component calls the ",(0,s.jsx)(n.code,{children:"click"})," prop after a\ndelay."]}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-js",children:"const Clicker = ({ click }) => (\n  <div>\n    <button onClick={() => setTimeout(click, 500)}>Click me</button>\n  </div>\n)\n"})}),"\n",(0,s.jsx)(t,{src:"/img/app/retry-ability/delay-click.png",alt:"Expect fails the test without waiting for the delayed stub"}),"\n",(0,s.jsxs)(n.p,{children:["The test finishes before the component calls the ",(0,s.jsx)(n.code,{children:"click"})," prop twice, and without\nretrying the assertion ",(0,s.jsx)(n.code,{children:"expect(onClick).to.be.calledTwice"}),"."]}),"\n",(0,s.jsx)(n.p,{children:"It could also fail if React or another JavaScript library decided to rerender\nthe DOM between clicks."}),"\n",(0,s.jsxs)(n.h4,{id:"Correctly-waiting-for-the-stub-to-be-called",children:[(0,s.jsx)(o,{name:"check-circle",color:"green"})," Correctly waiting for the stub to be called"]}),"\n",(0,s.jsxs)(n.p,{children:["We recommend aliasing the stub using the ",(0,s.jsx)(n.a,{href:"/api/commands/as",children:(0,s.jsx)(n.code,{children:".as"})})," command and\nusing ",(0,s.jsx)(n.code,{children:"cy.get('@alias')"})," to run assertions."]}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-js",children:"it('calls the click prop', () => {\n  const onClick = cy.stub().as('clicker')\n\n  cy.mount(<Clicker click={onClick} />)\n  // Good practice \ud83d\udca1: Don't chain anything off of commands\n  cy.get('button').click()\n  cy.get('button').click()\n\n  // Good practice \ud83d\udca1: Reference the stub with an alias\n  cy.get('@clicker').should('have.been.calledTwice')\n})\n"})}),"\n",(0,s.jsx)(t,{src:"/img/app/retry-ability/click-twice.gif",alt:"Retrying the assertions using a stub alias"}),"\n",(0,s.jsxs)(n.h3,{id:"Use-should-with-a-callback",children:["Use ",(0,s.jsx)(n.code,{children:".should()"})," with a callback"]}),"\n",(0,s.jsxs)(n.p,{children:["If you are using commands, but need to retry the entire chain, consider\nrewriting the commands into a\n",(0,s.jsx)(n.a,{href:"/api/commands/should#Function",children:".should(callbackFn)"}),"."]}),"\n",(0,s.jsx)(n.p,{children:"Below is an example where the number value is set after a delay:"}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-html",children:'<div class="random-number-example">\n  Random number: <span id="random-number">\ud83c\udf81</span>\n</div>\n<script>\n  const el = document.getElementById(\'random-number\')\n  setTimeout(() => {\n    el.innerText = Math.floor(Math.random() * 10 + 1)\n  }, 1500)\n<\/script>\n'})}),"\n",(0,s.jsx)(t,{src:"/img/app/retry-ability/random-number.gif",alt:"Random number"}),"\n",(0,s.jsxs)(n.h3,{id:"Incorrectly-waiting-for-values",children:[(0,s.jsx)(o,{name:"exclamation-triangle",color:"red"})," Incorrectly waiting for values"]}),"\n",(0,s.jsxs)(n.p,{children:["You may want to write a test like below, to test that the number is between 1\nand 10, although ",(0,s.jsx)(n.strong,{children:"this will not work as intended"}),". The test yields the\nfollowing values, noted in the comments, before failing."]}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-javascript",children:"// WRONG: this test will not work as intended\ncy.get('[data-testid=\"random-number\"]') // <div>\ud83c\udf81</div>\n  .invoke('text') // \"\ud83c\udf81\"\n  .then(parseFloat) // NaN\n  .should('be.gte', 1) // fails\n  .and('be.lte', 10) // never evaluates\n"})}),"\n",(0,s.jsxs)(n.p,{children:["Unfortunately, the ",(0,s.jsx)(n.a,{href:"/api/commands/then",children:".then()"})," command breaks the chain of\nqueries - nothing before it (such as fetching the text of the element) re-runs."]}),"\n",(0,s.jsx)(t,{src:"/img/app/retry-ability/random-number-first-attempt.png",alt:"First attempt at writing the test"}),"\n",(0,s.jsxs)(n.h4,{id:"Correctly-waiting-for-values",children:[(0,s.jsx)(o,{name:"check-circle",color:"green"})," Correctly waiting for values"]}),"\n",(0,s.jsxs)(n.p,{children:["We need to retry getting the element, invoking the ",(0,s.jsx)(n.code,{children:"text()"})," method, calling the\n",(0,s.jsx)(n.code,{children:"parseFloat"})," function and running the ",(0,s.jsx)(n.code,{children:"gte"})," and ",(0,s.jsx)(n.code,{children:"lte"})," assertions. We can achieve\nthis using the ",(0,s.jsx)(n.code,{children:".should(callbackFn)"}),"."]}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-javascript",children:"cy.get('[data-testid=\"random-number\"]').should(($div) => {\n  // all the code inside here will retry\n  // until it passes or times out\n  const n = parseFloat($div.text())\n\n  expect(n).to.be.gte(1).and.be.lte(10)\n})\n"})}),"\n",(0,s.jsxs)(n.p,{children:["The above test retries getting the element and invoking the text of the element\nto get the number. When the number is finally set in the application, then the\n",(0,s.jsx)(n.code,{children:"gte"})," and ",(0,s.jsx)(n.code,{children:"lte"})," assertions pass and the test passes."]}),"\n",(0,s.jsx)(t,{src:"/img/app/retry-ability/random-number-callback.gif",alt:"Random number using callback"}),"\n",(0,s.jsx)(n.h2,{id:"See-also",children:"See also"}),"\n",(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsxs)(n.li,{children:["Read our blog posts about fighting\n",(0,s.jsx)(n.a,{href:"https://cypress.io/blog/tag/flake/",children:"the test flake"}),"."]}),"\n",(0,s.jsxs)(n.li,{children:["You can add retry-ability to your own\n",(0,s.jsx)(n.a,{href:"/api/cypress-api/custom-commands",children:"custom commands"})," and queries."]}),"\n",(0,s.jsxs)(n.li,{children:["You can retry any function with attached assertions using the 3rd party\nplugins ",(0,s.jsx)(n.a,{href:"https://github.com/NicholasBoll/cypress-pipe",children:"cypress-pipe"})," and\n",(0,s.jsx)(n.a,{href:"https://github.com/NoriSte/cypress-wait-until",children:"cypress-wait-until"}),"."]}),"\n",(0,s.jsxs)(n.li,{children:["To learn how to enable Cypress' test retries functionality, which retries\ntests that fail, check out our official guide on\n",(0,s.jsx)(n.a,{href:"/app/guides/test-retries",children:"Test Retries"}),"."]}),"\n"]})]})}function h(e={}){const{wrapper:n}={...(0,i.a)(),...e.components};return n?(0,s.jsx)(n,{...e,children:(0,s.jsx)(d,{...e})}):d(e)}function u(e,n){throw new Error("Expected "+(n?"component":"object")+" `"+e+"` to be defined: you likely forgot to import, pass, or provide it.")}},1151:(e,n,t)=>{t.d(n,{Z:()=>r,a:()=>a});var s=t(7294);const i={},o=s.createContext(i);function a(e){const n=s.useContext(o);return s.useMemo((function(){return"function"==typeof e?e(n):{...n,...e}}),[n,e])}function r(e){let n;return n=e.disableParentContext?"function"==typeof e.components?e.components(i):e.components||i:a(e.components),s.createElement(o.Provider,{value:n},e.children)}}}]);